anni-provider/src/providers/priority.rs

```diff
@@ -0,0 +1,157 @@
+use std::{borrow::Cow, collections::HashSet, num::NonZeroU8};
+
+use async_trait::async_trait;
+
+use crate::{AnniProvider, AudioResourceReader, ProviderError, Range, ResourceReader, Result};
+
+#[derive(Default)]
+pub struct PriorityProvider(Vec<(i32, Box<dyn AnniProvider + Send + Sync>)>);
+
+impl PriorityProvider {
+    pub fn new(mut providers: Vec<(i32, Box<dyn AnniProvider + Send + Sync>)>) -> Self {
+        providers.sort_by(|(x, _), (y, _)| x.cmp(y).reverse());
+
+        Self(providers)
+    }
+
+    pub fn insert(&mut self, provider: Box<dyn AnniProvider + Send + Sync>, priority: i32) {
+        match self.0.binary_search_by(|(p, _)| p.cmp(&priority).reverse()) {
+            Ok(pos) | Err(pos) => self.0.insert(pos, (priority, provider)),
+        };
+    }
+}
+
+impl FromIterator<(i32, Box<dyn AnniProvider + Send + Sync>)> for PriorityProvider {
+    fn from_iter<T: IntoIterator<Item = (i32, Box<dyn AnniProvider + Send + Sync>)>>(
+        iter: T,
+    ) -> Self {
+        Self::new(iter.into_iter().collect())
+    }
+}
+
+impl<P: AnniProvider + Send + Sync + 'static> FromIterator<(i32, P)> for PriorityProvider {
+    fn from_iter<T: IntoIterator<Item = (i32, P)>>(iter: T) -> Self {
+        Self::new(
+            iter.into_iter()
+                .map(|(priority, provider)| (priority, Box::new(provider) as _))
+                .collect(),
+        )
+    }
+}
+
+#[async_trait]
+impl AnniProvider for PriorityProvider {
+    async fn albums(&self) -> Result<HashSet<Cow<str>>> {
+        let mut res = HashSet::new();
+
+        for (_, provider) in self.0.iter() {
+            res.extend(provider.albums().await?);
+        }
+
+        Ok(res)
+    }
+
+    async fn get_audio(
+        &self,
+        album_id: &str,
+        disc_id: NonZeroU8,
+        track_id: NonZeroU8,
+        range: Range,
+    ) -> Result<AudioResourceReader> {
+        for (_, provider) in self.0.iter() {
+            match provider.get_audio(album_id, disc_id, track_id, range).await {
```

Comment:
We may reuse some logic from `MultipleProviders`(or just replace `MultipleProviders` with this one)?

Reply:
The logic is relatively simple so I think there is no need.